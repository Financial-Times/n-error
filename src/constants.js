export const CATEGORIES = {
	FETCH_RESPONSE_OK: 'FETCH_RESPONSE_OK',
	FETCH_RESPONSE_ERROR: 'FETCH_RESPONSE_ERROR',
	FETCH_NETWORK_ERROR: 'FETCH_NETWORK_ERROR',
	NODE_SYSTEM_ERROR: 'NODE_SYSTEM_ERROR',
	CUSTOM_ERROR: 'CUSTOM_ERROR',
};

// HTTP status code
// http://www.restapitutorial.com/httpstatuscodes.html
export const ERROR_STATUS_TEXT_MAP = {
	'400': 'BAD_REQUEST',
	'401': 'UNAUTHORISED',
	// '402': 'PAYMENT_REQUIRED',
	'403': 'FORBIDDEN',
	'404': 'NOT_FOUND',
	// '405': 'METHOD_NOT_ALLOWED',
	// '406': 'NOT_ACCEPTABLE',
	// '407': 'PROXY_AUTHENTICATION_REQUIRED',
	'408': 'REQUEST_TIMEOUT',
	'409': 'CONFLICT',
	'500': 'INTERNAL_SERVER_ERROR',
	// '501': 'NOT_IMPLEMENTED',
	'502': 'BAD_GATEWAY',
	'503': 'SERVICE_UNAVAILABLE',
	'504': 'GATEWAY_TIMEOUT',
};

export default {
	CATEGORIES,
	ERROR_STATUS_TEXT_MAP,
};
