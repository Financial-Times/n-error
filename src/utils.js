export const assertError = e => {
	expect(e instanceof Error).toBe(true);
	const { message, stack, ...custom } = e;
	expect(stack).toBeDefined();
	expect({ message, ...custom }).toMatchSnapshot();
};

export const removeObjectKeys = obj => keys => {
	let output = {};
	if (Array.isArray(keys)) {
		Object.keys(obj).forEach(key => {
			const toBeRemoved = keys.includes(key);
			if (!toBeRemoved) {
				output[key] = obj[key];
			}
		});
	} else if (typeof keys === 'string' && keys !== '') {
		output = obj;
		delete output[keys];
	} else {
		throw Error('keys need to be formatted in [string] or non-empty string');
	}
	return output;
};
