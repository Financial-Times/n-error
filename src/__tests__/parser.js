import fetch from 'node-fetch';
import {
	parseFetchResponseError,
	parseFetchNetworkError,
	parseFetchError,
} from '../parser';
import { CATEGORIES } from '../constants';
import { assertError } from '../utils';

describe('parseFetchResponseError', () => {
	it('reports wrong response being threw to catch if response is ok', async () => {
		const headers = new Headers();
		headers.append('content-type', 'text/plain; charset=utf-8');
		const response = new Response('', { status: 200, headers });
		const parsed = await parseFetchResponseError(response);
		expect(parsed.category).toBe(CATEGORIES.FETCH_RESPONSE_OK);
		expect(parsed).toMatchSnapshot();
	});

	it('format error in text/html contentType correctly', async () => {
		const headers = new Headers();
		headers.append('content-type', 'text/html; charset=utf-8');
		const message = '<html></html>';
		const e = new Response(message, { status: 404, headers });
		const parsed = await parseFetchResponseError(e);
		expect(parsed.message).toBe(message);
		assertError(parsed);
	});

	it('format error in text/plain contentType correctly', async () => {
		const headers = new Headers();
		headers.append('content-type', 'text/plain; charset=utf-8');
		const message = '403 Forbidden';
		const e = new Response(message, { status: 403, headers });
		const parsed = await parseFetchResponseError(e);
		expect(parsed.message).toBe(message);
		assertError(parsed);
	});

	it('format error in application/json contentType correctly', async () => {
		const headers = new Headers();
		headers.append('content-type', 'application/json; charset=utf-8');
		const message = { message: 'some message', document: 'some url' };
		const e = new Response(JSON.stringify(message), { status: 404, headers });
		const parsed = await parseFetchResponseError(e);
		expect(parsed.message).toEqual(message);
		assertError(parsed);
	});
});

describe('parseFetchNetworkError', () => {
	it('format network error correctly', () => {
		const e = new fetch.FetchError('mocked fetch network error');
		e.code = 'MOCKED_FETCH_NETWORK_ERROR_CODE';
		const parsed = parseFetchNetworkError(e);
		expect(parsed.category).toBe(CATEGORIES.FETCH_NETWORK_ERROR);
		assertError(parsed);
	});
});

describe('parseFetchError', () => {
	it('format node-fetch response error correctly', async () => {
		const headers = new fetch.Headers();
		headers.append('content-type', 'text/plain; charset=utf-8');
		const e = new fetch.Response('403 Forbidden', { status: 403, headers });
		const parsed = await parseFetchError(e);
		expect(parsed.category).toBe(CATEGORIES.FETCH_RESPONSE_ERROR);
		assertError(parsed);
	});

	it('format isomorphic-fetch response error correctly', async () => {
		const headers = new Headers();
		headers.append('content-type', 'text/plain; charset=utf-8');
		const e = new Response('403 Forbidden', { status: 403, headers });
		const parsed = await parseFetchError(e);
		expect(parsed.category).toBe(CATEGORIES.FETCH_RESPONSE_ERROR);
		assertError(parsed);
	});

	it('format network error correctly', async () => {
		const e = new fetch.FetchError('mocked fetch network error');
		e.code = 'MOCKED_FETCH_NETWORK_ERROR_CODE';
		const parsed = await parseFetchError(e);
		expect(parsed.category).toBe(CATEGORIES.FETCH_NETWORK_ERROR);
		assertError(parsed);
	});

	it('does nothing on other error types e.g. System Error ', async () => {
		const e = new Error('some error message');
		const parsed = await parseFetchError(e);
		expect(parsed).toBe(e);
	});
});
