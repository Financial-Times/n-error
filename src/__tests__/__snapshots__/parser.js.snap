// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseFetchError format isomorphic-fetch response error correctly 1`] = `
Object {
  "category": "FETCH_RESPONSE_ERROR",
  "contentType": "text/plain; charset=utf-8",
  "extend": [Function],
  "message": "403 Forbidden",
  "remove": [Function],
  "setHandler": [Function],
  "status": 403,
  "toUser": [Function],
}
`;

exports[`parseFetchError format network error correctly 1`] = `
Object {
  "category": "FETCH_NETWORK_ERROR",
  "code": "MOCKED_FETCH_NETWORK_ERROR_CODE",
  "extend": [Function],
  "message": "mocked fetch network error",
  "remove": [Function],
  "setHandler": [Function],
  "toUser": [Function],
}
`;

exports[`parseFetchError format node-fetch response error correctly 1`] = `
Object {
  "category": "FETCH_RESPONSE_ERROR",
  "contentType": "text/plain; charset=utf-8",
  "extend": [Function],
  "message": "403 Forbidden",
  "remove": [Function],
  "setHandler": [Function],
  "status": 403,
  "toUser": [Function],
}
`;

exports[`parseFetchNetworkError format network error correctly 1`] = `
Object {
  "category": "FETCH_NETWORK_ERROR",
  "code": "MOCKED_FETCH_NETWORK_ERROR_CODE",
  "extend": [Function],
  "message": "mocked fetch network error",
  "remove": [Function],
  "setHandler": [Function],
  "toUser": [Function],
}
`;

exports[`parseFetchResponseError format error in application/json contentType correctly 1`] = `
Object {
  "category": "FETCH_RESPONSE_ERROR",
  "contentType": "application/json; charset=utf-8",
  "extend": [Function],
  "message": Object {
    "document": "some url",
    "message": "some message",
  },
  "remove": [Function],
  "setHandler": [Function],
  "status": 404,
  "toUser": [Function],
}
`;

exports[`parseFetchResponseError format error in text/html contentType correctly 1`] = `
Object {
  "category": "FETCH_RESPONSE_ERROR",
  "contentType": "text/html; charset=utf-8",
  "extend": [Function],
  "message": "<html></html>",
  "remove": [Function],
  "setHandler": [Function],
  "status": 404,
  "toUser": [Function],
}
`;

exports[`parseFetchResponseError format error in text/plain contentType correctly 1`] = `
Object {
  "category": "FETCH_RESPONSE_ERROR",
  "contentType": "text/plain; charset=utf-8",
  "extend": [Function],
  "message": "403 Forbidden",
  "remove": [Function],
  "setHandler": [Function],
  "status": 403,
  "toUser": [Function],
}
`;

exports[`parseFetchResponseError reports wrong response being threw to catch if response is ok 1`] = `
Object {
  "category": "FETCH_RESPONSE_OK",
  "message": "it shouldn't be caught as exception, please check the code",
  "status": 200,
}
`;
